### ISSUE TITLE
The full-screen button on videos does not work as expected

### ISSUE THREAD

**Issue by tito72** [QupZilla/qupzilla#2233]
When the window is maximized and you click on the button at the bottom right of a HTML5 video, Qupzilla does not go to full screen contrary to that when the window is not maximized.

**Comment by nowrep** [comment]
Can you please provide some testing page? Also what is your QupZilla version?

**Comment by tito72** [comment]
I tested it with a video open directly in Qupzilla. I use the latest version compiled from this Github repository with ArchLinux AUR. I also forget to say that I am using Plasma 5.9.2.

**Comment by nowrep** [comment]
I can't reproduce it, it correctly switches to fullscreen for me regardless of window being maximized or not.
Can you please create screencast with the issue?

**Comment by tito72** [comment]
When I open the video it is like that: 
![screenshot_1](https://cloud.githubusercontent.com/assets/3169588/23334940/23517d8e-fbaa-11e6-83a0-0f1da13010e1.png)
Then when I click on the fullscreen button it is like that:
![screenshot_2](https://cloud.githubusercontent.com/assets/3169588/23334942/2483cb44-fbaa-11e6-9b1b-048ebcb4598d.png)
The third time I click on the button it is that time in fullscreen. The fourth time I click it is like that:
![screenshot_3](https://cloud.githubusercontent.com/assets/3169588/23334965/9e741c42-fbaa-11e6-9cb2-7c196daec0f8.png)
I also noticed that this happens only when the video is playing.

**Comment by cranes-bill** [comment]
I'm also facing that behaviour from time to time with Qt 5.6, but it occurs only the first time i click _maximize_ - _minimize_, then works normally.

**Comment by cranes-bill** [comment]
I'm guessing the following Chromium output is related:
`requestFullscreen() is deprecated on insecure origins, and support will be removed in the future. You should consider switching your application to a secure origin, such as HTTPS. See https://goo.gl/rStTGz for more details.`

**Comment by MCMic** [comment]
Same problem here on youtube and twitch, using qupzilla-git from AUR on ArchLinux

**Comment by cranes-bill** [comment]
Lets see if can give some clue here.
I noticed that when this occurs, the whole browser was switched on to full-screen instead of video only _(the same as F11 clicked)_. Also noticed that shading the window before watching the video may force triggering this behavior.
I tried to make sample [video](https://yadi.sk/i/IwiKfIMp3RAqQQ), but the player is not able to capture the cursor, so it might not be so useful.

**Comment by richarson** [comment]
I have a similar bug that I can easily repoduce on youtube, mainly the part of leaving (JS-based?) fullscreen from youtube un-maximizing Qupzilla's window.

I have Qupzilla 2.2.2. compiled by me against Qt 5.9.2, I can try patches if needed.

**Comment by Diablopc** [comment]
> @richarson Qupzilla 2.2.2

There is a lot of work were done since that time including fullscreen handling.
Try build qupzilla from git.

**Comment by richarson** [comment]
I was waiting for 2.2.4 to be released, but I cant try with git.

**Comment by cranes-bill** [comment]
@richarson That issue is not resolved yet _(otherwise it should be closed)_ and probably will not be fixed in 2.2.4.

**Comment by nowrep** [comment]
I can't still reproduce it, but yeah, please test with 2.2.4 when it's out.

**Comment by Diablopc** [comment]
i've tried it few times but still can't reproduce

**Comment by cranes-bill** [comment]
@nowrep I already tested it with 2.2.4 :smile: and the issue persists.

**Comment by Diablopc** [comment]
@cranes-bill Can you describe proper algorithm to reproduce?

**Comment by Diablopc** [comment]
Btw
> @nowrep with 2.2.4 when it's out.
When it must happen?

**Comment by cranes-bill** [comment]
> Can you describe proper algorithm to reproduce?


Just toggle fullscreen. Also see my [comment](https://github.com/QupZilla/qupzilla/issues/2233#issuecomment-354987013) above. But it is not 100% reproducible. Somtimes it occours 2 times of 10, somtimes 8/10

**Comment by cranes-bill** [comment]
> with 2.2.4 when it's out. When it must happen?


Oh yes, when the version digits will be changed, everything is go away. ;D

**Comment by nowrep** [comment]
@cranes-bill No issues with normal fullscreen? (F11 or action in menu)

**Comment by cranes-bill** [comment]
> No issues with normal fullscreen? (F11 or action in menu)

**I think no.**
I also was facing that in some image viewers where have a full screen button.
Also i think you be able to reproduce it on any „non-KDE“ system.

**Comment by cranes-bill** [comment]
@nowrep Actualli i enabled the „Full screen button“ and also reproduced it.

**Comment by nowrep** [comment]
What fullscreen button?

**Comment by cranes-bill** [comment]
In toolbar

**Comment by nowrep** [comment]
But that can only be used to exit fullscreen, not enter.

**Comment by cranes-bill** [comment]
Yes, i entered with F11, then i hit this button...
But now it got reproducible with F11 only.

**Comment by nowrep** [comment]
Hmm, works fine here even with Openbox.

**Comment by cranes-bill** [comment]
OK, the button does not matter, it is reproducible with F11 as well.

> No issues with normal fullscreen? (F11 or action in menu)

So the right answer is - Also has an issue there

> Hmm, works fine here even with Openbox.

Only left to be hardware related..?

**Comment by nowrep** [comment]
No, it will most likely be race between Qt and X server. I assume that it will enter fullscreen and then immediately leave it for some reason.

Can you please post output when it happens?

```diffdiff --git a/src/lib/app/browserwindow.cpp b/src/lib/app/browserwindow.cpp
index 98927be3..4b377782 100644
--- a/src/lib/app/browserwindow.cpp
+++ b/src/lib/app/browserwindow.cpp
@@ -1195,6 +1195,7 @@ bool BrowserWindow::event(QEvent *event)
 {
     if (event->type() == QEvent::WindowStateChange) {
         QWindowStateChangeEvent *e = static_cast<QWindowStateChangeEvent*>(event);
+        qDebug() << "old" << e->oldState() << "now" << windowState();
         if (!(e->oldState() & Qt::WindowFullScreen) && windowState() & Qt::WindowFullScreen) {
             // Enter fullscreen
             m_statusBarVisible = statusBar()->isVisible();
```

**Comment by cranes-bill** [comment]
May i try that with Falkon?

**Comment by nowrep** [comment]
Yes

**Comment by cranes-bill** [comment]
@nowrep I can't get anything on the output, what i'm missing!? 
(built with `-DCMAKE_BUILD_TYPE=Debug` )

**Comment by nowrep** [comment]
You are probably loading old library, as usual :D

**Comment by cranes-bill** [comment]
I also built QupZilla, but there is no `qDebug` flags in the library by any reason.
I will build an AppImage with this code, so please be patient!

**Comment by cranes-bill** [comment]
OK, i built QupZilla with precompiled Qt 5.10 libraries and now have debug output, but fail to reproduce the problem for now, so it's probably in Qt 5.9
I will still try...

**Comment by cranes-bill** [comment]
Well, i built QupZilla with precompiled Qt 5.9.2 and reproduced the problem.
Posting the last lines right after it occoured. But since i fail to reproduce it with Qt 5.10, it is pretty clear where is it.....
```
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen)
old QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowNoState)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen)
old QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowNoState)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen)
old QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowNoState)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized)
old QFlags<Qt::WindowState>(WindowMaximized) now QFlags<Qt::WindowState>(WindowMaximized|WindowFullScreen)
old QFlags<Qt::WindowState>(WindowFullScreen) now QFlags<Qt::WindowState>(WindowNoState)
old QFlags<Qt::WindowState>(WindowNoState) now QFlags<Qt::WindowState>(WindowFullScreen)
old QFlags<Qt::WindowState>(WindowFullScreen) now QFlags<Qt::WindowState>(WindowNoState)

```
